### PR TITLE
Ignore package-lock.json files when validating

### PR DESCRIPTION
**Pull request by alexkilgour**
Fixes a small bug when validating packages on the command line.

When running tests for a package within a toolkit, you need to install any dependencies before running the tests (this can be done with [this command](https://github.com/springernature/frontend-toolkits/blob/master/package.json#L22)). This creates a `package-lock.json` file within the package.

We [don't want to commit these files](https://www.twilio.com/blog/lockfiles-nodejs), so we exclude them in our [`.gitignore` file](https://github.com/springernature/frontend-toolkits/blob/master/.gitignore#L67).

When validating our packages, the package manager parses that `.gitignore` file and does not run the validator on the filepaths generated.

There was a bug where the package level lock files were still being validated (and throwing errors) because the current working directory changes when validating packages. This also caused problems in testing because of the way the filesystem gets mocked in Jest.

This update:
* Refactors the code to fix this issue and also allows for better, more robust testing.
* Refactors some of the variable naming to better explain their function

### CODE REVIEW

File: __mocks__/gitignore-globs.js [hunk 1]
```diff
@@ -5,5 +5,5 @@
 'use strict';
 
 module.exports = () => {
-	return ['folder3'];
+	return ['**/folder3', '**/folder3/**', '**/packages/**/package-lock.json'];
```
Comment (alexkilgour) [inline __mocks__/gitignore-globs.js]: This change better represents how the `.gitignore` files are parsed

File: __mocks__/fs.js [hunk 2]
```diff
@@ -36,6 +36,7 @@ const defaultPackageContents = {
 const __fsMockFiles = () => {
 	return {
 		'packages/package/pass': defaultPackageContents,
+		'packages/package/passWithBuildFiles': {...defaultPackageContents, ...{'package-lock.json': 'file content'}},
```
Comment (alexkilgour) [inline __mocks__/fs.js]: test that lock files in packages are ignored

File: __tests__/unit/_validate/check-package-structure.test.js [hunk 3]
```diff
@@ -56,6 +56,13 @@ describe('Check validation', () => {
 		).resolves.toEqual();
 	});
 
+	test('Resolves when filesystem matches config ignore build files', async () => {
```
Comment (jpw) [inline __tests__/unit/_validate/check-package-structure.test.js]: eh?

Comment (alexkilgour) [reply __tests__/unit/_validate/check-package-structure.test.js]: 🤔 yeah, could probably phrase that better

Comment (alexkilgour) [reply __tests__/unit/_validate/check-package-structure.test.js]: updated

**Comment by benjclark** [comment]
Interesting article about package-lock being ignored in packages published to npm.

For me it raised a couple of questions...

1. should we add `package-lock=false` to `.npmrc` in the toolkit repo so to ensure package-lock files are never generated in the first place?

2. should we `npm shrinkwrap` our packages to get the same benefits of package-lock files?

File: lib/js/_validate/_check-package-structure.js [hunk 4]
```diff
@@ -17,26 +17,44 @@ const getPackageName = require('../_utils/_get-package-name');
...
+ * Check if a file/folder matches any files excluded from validation
+ * Allows partial matches on the end of the path to cater for cwd
+ * Returns true if match found
+ * @private
+ * @function filterExcludedFiles
+ * @param {String} localPath path to a file or folder to check
+ * @param {Array} excludedFiles Array of file paths to exclude from validation
+ * @return {Boolean}
+ */
+function filterExcludedFiles(localPath, excludedFiles) {
```
Comment (jpw) [inline lib/js/_validate/_check-package-structure.js]: `function pathIsExcluded(localPath, excludedFiles) => excludedFiles.some(excludedItem => localPath.endsWith(excludedItem));`
maybe too terse, but will return early

File: lib/js/_validate/_check-package-structure.js [hunk 5]
```diff
@@ -116,34 +134,37 @@ function isFileType(item) {
...
  * - Remove paths matching .gitignore'd files
  * - Remove dotfiles
  * @private
  * @function getFilteredResults
- * @param {String} directory path to files
  * @param {Array} files file list to filter
  * @return {Array}
  */
-function getFilteredResults(directory, files) {
+function getFilteredResults(files) {
```
Comment (jpw) [inline lib/js/_validate/_check-package-structure.js]: Bit ambiguous..? maybe `removeIgnoredFiles`? or split into two functions, `removeDotFiles` and `removeGitIgnoredFiles` ?

File: lib/js/_validate/_check-package-structure.js [hunk 6]
```diff
@@ -116,34 +134,37 @@ function isFileType(item) {
...
-				matchBase: true,
-				dot: true
-			}
-		)
-	);
 
-	// https://github.com/regexhq/dotfile-regex/blob/master/index.js
-	const dotfileRegex = /(?:^|[\\\/])(\.(?!\.)[^\\\/]+)$/; // eslint-disable-line no-useless-escape
+	// Generate a pattern matching string from parsed .gitignore
+	const pattern = (globs.length > 1) ? `{${globs.join()}}` : globs[0];
```
Comment (jpw) [inline lib/js/_validate/_check-package-structure.js]: const pattern = \`{${globs.join()}}\`;
?

**Comment by alexkilgour** [comment]
> 1. should we add `package-lock=false` to `.npmrc` in the toolkit repo so to ensure package-lock files are never generated in the first place?
> 2. should we `npm shrinkwrap` our packages to get the same benefits of package-lock files?

1. I thought about this but not sure we can, because we still want to lock file for the repo, just not the individual packages. We use the main lock file to make sure everyone is running the same test/linting setup.

2. Possibly. Need to think through the ramifications e.g. If we use semver properly then by using the `^` selector we should get minor and patch updates without having to update the versions in our package.json file (create a new PR etc). Would this stop that behaviour? is that desireable?

File: lib/js/_validate/_check-package-structure.js (hunk 6 above)
Comment (alexkilgour) [reply lib/js/_validate/_check-package-structure.js]: that is how a pattern is represented in minimatch, as a string of the form `{matcha,matchb,matchc}`

File: lib/js/_validate/_check-package-structure.js (hunk 5 above)
Comment (alexkilgour) [reply lib/js/_validate/_check-package-structure.js]: It's explained in the function comment, but can rename it. not sure splitting it out serves to make it any clearer.

File: lib/js/_validate/_check-package-structure.js (hunk 6 above)
Comment (alexkilgour) [reply lib/js/_validate/_check-package-structure.js]: i'll add an example to the comment

File: lib/js/_validate/_check-package-structure.js (hunk 4 above)
Comment (alexkilgour) [reply lib/js/_validate/_check-package-structure.js]: oh nice 👍

**Comment by benjclark** [comment]
re 2

Yeah I suppose that wouldn't be great, it's best to allow consumers to benefit from the version range caret. Also, that article and npm recommend using it on CLIs and devDependencies but not libraries which makes sense as minor/patch updates to libraries generally means end user will get those benefits which is important for things like security. This repo is essentially a CLI, right? Perhaps shrinkwrap it could be of use in this repo specifically to ensure the tree of child dependencies is consistent for every consumer?

File: lib/js/_validate/_check-package-structure.js (hunk 6 above)
Comment (jpw) [reply lib/js/_validate/_check-package-structure.js]: Ahhh, I was thinking the conditional was unnecessary but I think i get it now. You are saying the pattern cannot be `{foo}` i.e. if one entry, it must be a string?

**Comment by alexkilgour** [comment]
You mean shrinkwrap for the repo itself, not the packages? We already use the `package-lock.json` file for the repo, and you don't need both

> The new lockfile ("package-lock.json") shares basically all of the same code, the exact same format as npm-shrinkwrap (you can rename them between one another!). It's also something the community seems to understand: "it has a lockfile" seems to click so much faster with people. Finally, having a new file meant that we could have relatively low-risk backwards-compat with shrinkwrap without having to do weird things like allow-publication mentioned in the parent post.

File: lib/js/_validate/_check-package-structure.js (hunk 6 above)
Comment (alexkilgour) [reply lib/js/_validate/_check-package-structure.js]: yes!

**Comment by alexkilgour** [comment]
> This repo is essentially a CLI, right? Perhaps shrinkwrap it could be of use in this repo specifically to ensure the tree of child dependencies is consistent for every consumer?

As discussed in person, what we really need is that every consumer _of the same type_ is consistent. So in our use case, we use it for the `frontend toolkits` repo which enforces consistency through the use of a `package-lock.json` file within that repo. This means every instance of `frontend toolkits` can have the same tree. If this gets used somewhere else then that consumer can enforce consistency using the `package-lock.json` file, but we don't necessarily need consistency between the two different types of consumer.

**Pull request merged**